Image recognition example for Tensorflow Lite Micro

jenselofsson opened the pull request:
This is a demo showing how to use Tensorflow Lite Micro to perform image recognition using a STM32F746G-DISCO discovery board with a STM32F4DIS-CAM camera module attached.

jenselofsson committed:
Don't build the RTOS component of Mbed OS in order to save flash/RAM
memory.

jenselofsson committed:
Add the micro_recognition example.

Image recognition on the DISCO board using the CIFAR10 dataset.

jenselofsson committed:
Add micro_recognition testcase.

This test download the cifar10 dataset, extracts the first 10 images
from the test_batch.bin into fira_10_cifar_images.cc and moves it into
the micro_recognition folder. In order for the test to pass 6/10 images
must be correctly guessed.

jenselofsson committed:
Add a README file for the micro_recognition example.

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/disco/display_util.cc:
```diff
@@ -0,0 +1,79 @@
+
+void init_lcd() { lcd.Clear(LCD_COLOR_WHITE); }
+
+void display_image_rgb888(int x_dim, int y_dim, uint8_t* image_data, int x_loc,
```
const uint8_t* image_data

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/disco/display_util.cc:
```diff
@@ -0,0 +1,79 @@
+                          int y_loc) {
+  for (int y = 0; y < y_dim; y++) {
+    for (int x = 0; x < x_dim; x++) {
+      int pix_loc = (y * x_dim + x) * 3;
```
I'd suggest removing `pix_loc`.
```cpp
for (int y = 0; y < y_dim; ++y) {
  for (int x = 0; x < x_dim; ++x, image_data += 3) {
     auto r = image_data[0];
     auto g = image_data[1];
     auto b = image_data[2];
     .....
  }
}
```

In addition, `++y` and `++x` is preferred over `y++` and `x++`.

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/disco/display_util.cc:
```diff
@@ -0,0 +1,79 @@
+                          int y_loc) {
+  for (int y = 0; y < y_dim; y++) {
+    for (int x = 0; x < x_dim; x++) {
+      int pix_loc = (y * x_dim + x) * 2;
```
refer to above to remove `pix_loc`
and  change
`image_data` to `const uint8_t*`.

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/disco/image_provider.cc:
```diff
@@ -0,0 +1,33 @@
+
+TfLiteStatus GetImage(tflite::ErrorReporter* error_reporter, int frame_width,
+                      int frame_height, int channels, uint8_t* frame) {
+  BSP_CAMERA_SnapshotStart(frame);
```
only the parameter `frame` is used here and why do you need to declare the other 3 parameters?

jenselofsson replied:
I used the GetImage() function from the micro speech example as a template for this function, since I'd thought it would be good to be consistent across the examples. But I'll (void) the unused parameters to avoid compilation warning, and I'll add a comment explaining why some parameters are unused.

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/disco/image_util.cc:
```diff
@@ -0,0 +1,49 @@
+void ResizeConvertImage(tflite::ErrorReporter* error_reporter,
+                        int in_frame_width, int in_frame_height,
+                        int num_in_channels, int out_frame_width,
+                        int out_frame_height, int channels, uint8_t* in_image,
```
const uint8_t* in_image,

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/disco/image_util.cc:
```diff
@@ -0,0 +1,49 @@
+  int width_offset = ((in_frame_width - in_frame_height) / 2) * num_in_channels;
+
+  int yresize_ratio = (in_frame_height / out_frame_height) * num_in_channels;
+  int xresize_ratio = (in_frame_width / out_frame_width) * num_in_channels;
```
you get `int` if you use `int / int * int` and there exist some numerical problems.
is this what you expect?

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/main.cc:
```diff
@@ -0,0 +1,101 @@
+  tflite::ErrorReporter* error_reporter = &micro_error_reporter;
+
+  if (InitCamera(error_reporter) != kTfLiteOk) {
+    error_reporter->Report("Failed to init camera");
```
```
error_reporter->Report("Failed to init camera\n");
```

`\n` is missing.

jenselofsson replied:
\n isn't neccessary, since DebugLogPrintf() (called by error_reporter->Report()) appends \r\n.
https://github.com/tensorflow/tensorflow/blob/47b46048a97444b332bb7bd36b5097676e2c489d/tensorflow/lite/experimental/micro/micro_error_reporter.cc#L59

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/main.cc:
```diff
@@ -0,0 +1,101 @@
+
+  tflite::ops::micro::AllOpsResolver resolver;
+
+  const int tensor_arena_size = 63 * 1024;
```
```cpp
constexpr int tensor_arena_size = 63 * 1024;
```

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/main.cc:
```diff
@@ -0,0 +1,101 @@
+                       input->data.uint8);
+
+    if (input->type != kTfLiteUInt8) {
+      error_reporter->Report("Wrong input type");
```
`\n` is missing

jenselofsson replied:
See reply on the previous comment regarding Report().

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/util.h:
```diff
@@ -0,0 +1,63 @@
+#define G_CHANNEL_OFFSET CHANNEL_SIZE
+#define B_CHANNEL_OFFSET (CHANNEL_SIZE * 2)
+
+int get_top_prediction(uint8_t* predictions) {
```
```cpp
const uint8_t* predictions
```

csukuangfj commented on tensorflow/lite/experimental/micro/examples/micro_recognition/util.h:
```diff
@@ -0,0 +1,63 @@
+  int max_score = 0;
+  int guess = 0;
+
+  for (int category_index = 0; category_index < 10; category_index++) {
```
should `10` be passed as an argument?

I would suggest assigning `predications[0]` to `max_score` first and then starts from index `1`.

jenselofsson committed:
Fix review comments.

petewarden commented:
Sorry for the slow response. Overall this looks great, thanks! We will have to move the model file outside of the repo unfortunately, since it's 280KB and larger files like that cause problems as they grow the download for all users.

You can see how we do something similar for the person detection model here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/experimental/micro/examples/micro_vision/Makefile.inc#L1

It would also be great if you could update the readme to mention the specific kinds of recognition that this model is doing, "Plane", "Car",  "Bird", "Cat", "Deer",  "Dog",  "Frog", "Horse", "Ship",  "Truck", and that it will always try to put what it's shown into one of these categories. Could you also rename it to cifar_recognition? I realize I started a confusing trend with "micro_speech", I should have named it something clearer.

jenselofsson commented:
> Sorry for the slow response. Overall this looks great, thanks! We will have to move the model file outside of the repo unfortunately, since it's 280KB and larger files like that cause problems as they grow the download for all users.
> 
> You can see how we do something similar for the person detection model here:
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/experimental/micro/examples/micro_vision/Makefile.inc#L1
> 

Sure, I'll take a look at that.

> It would also be great if you could update the readme to mention the specific kinds of recognition that this model is doing, "Plane", "Car", "Bird", "Cat", "Deer", "Dog", "Frog", "Horse", "Ship", "Truck", and that it will always try to put what it's shown into one of these categories. 

That's a good idea, reading through the README I realize that it's unclear on what the example actually does.

> Could you also rename it to cifar_recognition? I realize I started a confusing trend with "micro_speech", I should have named it something clearer.

Yes, no problem.

jenselofsson committed:
Fix review comments.

Change all occurences of micro_recognition to image_recognition.
Added a line to the README.md explaining what the application does, and
that a camera module is used.
Added a .gitignore file and moved the first_10_cifar_images.h from
the .gitignore in the root directory to the newly created one.
Removed unnecessary python command from build instructions.

jenselofsson committed:
Remove the model from the repo.

Removed the model from the repo and remove the image recognition test
from the Makefile.inc until the model is available.
Also included a bugfix to image_recognition_test.cc where a function
parameter was forgotten.

jenselofsson commented:
I've updated the PR with regards to the review comments. I've removed the testcase from the Makefile.inc in order to prevent it from running until the model is made available for download, something which I'm working on.

The commits should be squashed before merging, since the model appears in the git-history twice, one when adding and one when removing. If it's not done automatically, I can do it myself and force push after the changes have been approved.

jenselofsson committed:
Download the image recognition model when building.

jenselofsson committed:
Merge branch 'master' into micro_recognition

jenselofsson commented:
@petewarden  I've added the URL for the network model, and it builds fine. And I've come to think of one thing.

Since the testcase (image_recognition_test.cc) uses the first 10 images in the CIFAR10 dataset of test images, I've added so that the CIFAR10 image set is downloaded. I've only managed to find it distributed as a single tar.gz-file including all of the 60000 images, and it is quite large (~160 MB).

So it could take a while on a slow internet connection. I tried looking through the Makefile to see if there is a way do only download the image set if the image_recognition_test is being built.

I wasn't able to really find a solution, since it seems like the Makefile.inc:s in micro/examples/ are called before it is determined which exact testcases are to be built.

Do you (or anyone) know if this is a problem in the first case, and if it can be solved without making large changes to the make system?

jenselofsson commented:
@petewarden gentle reminder.

jenselofsson committed:
Merge remote-tracking branch 'upstream/master' into micro_recognition

jenselofsson committed:
Adapt to the move of micro.

Remove "experimental" from paths and include guards due to the
micro-directory moving from tensorflow/lite/experimental to
tensorflow/lite.

The URL/md5sum to the zip-file containing the model have also been
updated.

gbaned commented:
@jenselofsson Can you please resolve conflicts? Thanks!

jenselofsson commented:
@gbaned Done!

jenselofsson commented:
The "Resolve conflicts" button was greyed out, so I merge the upstream/master-branch locally. Let me know if I need to fix it further in any way.

gbaned commented:
@jenselofsson Can you please check reviewer comments and keep us posted. Thanks!

tensorflowbutler commented:
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

jenselofsson commented:
@tensorflowbutler It's being worked on.

jenselofsson committed:
Micro: Remove IMAGE_ARRAY define.

Remove the IMAGE_ARRAY define as per review comments and use the
original variable name.

jenselofsson commented on tensorflow/lite/experimental/micro/examples/micro_recognition/micro_recognition_test.cc:
```diff
@@ -0,0 +1,94 @@
+
+#include "tensorflow/lite/experimental/micro/examples/micro_recognition/first_10_cifar_images.h"
+
+#define IMAGE_ARRAY \
```
Fixed. I just removed the IMAGE_ARRAY define and using the variable name directly.

jenselofsson committed:
Merge branch 'master' into micro_recognition

jenselofsson commented:
When I try to view the "MacOS CPU Python3" build logs, it shows a 404 error.

jenselofsson committed:
Merge branch 'master' into micro_recognition

jenselofsson commented:
@njeffrie @petewarden gentle ping for review

jenselofsson committed:
Merge branch 'master' into micro_recognition

jenselofsson committed:
Merge branch 'master' into micro_recognition

jenselofsson commented:
@petewarden Ping for review.

jenselofsson committed:
Move image_recognition to examples/experimental/.

petewarden reviewed (changes requested):
Can we move this into a new tensorflow/lite/examples/experimental folder? Otherwise looks good to go, thanks!

gbaned commented:
@jenselofsson Can you please check petewarden's comments and keep us posted? Thanks!

jenselofsson commented:
@petewarden No problem, updating the PR ASAP.

jenselofsson committed:
Add wildcard for tensorflow/lite/micro/examples/experimental/-folder.

jenselofsson committed:
Merge branch 'master' into micro_recognition

jenselofsson commented:
@petewarden Ping for review.

njeffrie commented:
Overall looks good.  If possible, could you rename the "disco" directory to something a bit more descriptive?  Perhaps stm32f746_discovery or stm32f746_disco.

njeffrie reviewed (dismissed):
Other than "disco" directory rename this looks good.  Thanks again for the hard work and apologies for the delay.

jenselofsson committed:
Rename disco/ to stm32f746_discovery/.

jenselofsson commented:
@njeffrie Done!

rthadur commented on tensorflow/lite/micro/examples/experimental/image_recognition/image_recognition_test.cc:
```diff
@@ -0,0 +1,95 @@
+
+  const tflite::Model* model = ::tflite::GetModel(image_recognition_model_data);
+  if (model->version() != TFLITE_SCHEMA_VERSION) {
+    error_reporter->Report(
```
`TF_LITE_REPORT_ERROR should be used instead of error_reporter, so log strings can be removed to save space if needed.`

rthadur commented on tensorflow/lite/micro/examples/experimental/image_recognition/image_recognition_test.cc:
```diff
@@ -0,0 +1,95 @@
+    TfLiteStatus invoke_status = interpreter.Invoke();
+    TF_LITE_MICRO_EXPECT_EQ(kTfLiteOk, invoke_status);
+    if (invoke_status != kTfLiteOk) {
+      error_reporter->Report("Invoke failed\n");
```
`TF_LITE_REPORT_ERROR should be used instead of error_reporter, so log strings can be removed to save space if needed.`

rthadur commented on tensorflow/lite/micro/examples/experimental/image_recognition/main.cc:
```diff
@@ -0,0 +1,101 @@
+  init_lcd();
+  wait_ms(100);
+
+  tflite::MicroErrorReporter micro_error_reporter;
```
`TF_LITE_REPORT_ERROR should be used instead of error_reporter, so log strings can be removed to save space if needed.`

rthadur commented on tensorflow/lite/micro/examples/experimental/image_recognition/main.cc:
```diff
@@ -0,0 +1,101 @@
+  if (InitCamera(error_reporter) != kTfLiteOk) {
+    error_reporter->Report("Failed to init camera.");
+    return 1;
+  }
```
`TF_LITE_REPORT_ERROR should be used instead of error_reporter, so log strings can be removed to save space if needed.`

rthadur commented on tensorflow/lite/micro/examples/experimental/image_recognition/main.cc:
```diff
@@ -0,0 +1,101 @@
+             camera_buffer);
+
+    ResizeConvertImage(error_reporter, IN_IMG_WIDTH, IN_IMG_HEIGHT, NUM_IN_CH,
+                       CNN_IMG_SIZE, CNN_IMG_SIZE, NUM_OUT_CH, camera_buffer,
```
`TF_LITE_REPORT_ERROR should be used instead of error_reporter, so log strings can be removed to save space if needed.`

rthadur commented on tensorflow/lite/micro/examples/experimental/image_recognition/main.cc:
```diff
@@ -0,0 +1,101 @@
+
+    if (input->type != kTfLiteUInt8) {
+      error_reporter->Report("Wrong input type.");
+    }
```
`TF_LITE_REPORT_ERROR should be used instead of error_reporter, so log strings can be removed to save space if needed.`

rthadur commented on tensorflow/lite/micro/examples/experimental/image_recognition/stm32f746_discovery/image_provider.cc:
```diff
@@ -0,0 +1,39 @@
+
+TfLiteStatus InitCamera(tflite::ErrorReporter* error_reporter) {
+  if (BSP_CAMERA_Init(RESOLUTION_R160x120) != CAMERA_OK) {
+    error_reporter->Report("Failed to init camera.\n");
```
`TF_LITE_REPORT_ERROR should be used instead of error_reporter, so log strings can be removed to save space if needed.`

rthadur reviewed (changes requested):
can you please below errors which were reported internally ?

jenselofsson committed:
Replace Report() with TF_LITE_REPORT_ERROR-macro.

jenselofsson commented:
@rthadur Done!

jenselofsson commented:
The error in the Windows Bazel build seems to be due to keras, which is unrelated to this PR.

freddan80 commented:
@jenselofsson, @rthadur is it ready to merge?

rthadur commented:
@njeffrie can you please help merge this PR internally ?

jenselofsson commented:
@rthadur @njeffrie Ping for review

rthadur commented:
@njeffrie is working internally to merge this PR , lets wait until he comes back.Thanks for your patience.
